File: dsa-23au/java-dsa/pswish-pipeline/scratch/Coins.java
package main.java.com.pswishcorp.app;

import java.util.HashMap;
import java.util.Map;

public class MakeChangeMinCoins {

    private static Map<Integer, Integer> memo = new HashMap<>();

    public static int makeChangeMinCoins(int[] coins, int amount) {
        // Sort coins in descending order
        Arrays.sort(coins);

        // If amount is 0, return 0
        if (amount == 0) {
            return 0;
        }

        // If amount is less than 0 or there are no coins, return infinity
        if (amount < 0 || coins.length == 0) {
            return Integer.MAX_VALUE;
        }

        // Check if the current subproblem has already been solved
        if (memo.containsKey(amount)) {
            return memo.get(amount);
        }

        // Calculate the minimum number of coins to make the amount, including the last coin
        int includeLastCoin = 1 + makeChangeMinCoins(coins, amount - coins[0]);

        // Calculate the minimum number of coins to make the amount, excluding the last coin
        int excludeLastCoin = makeChangeMinCoins(coins, amount, coins[1]);

        // Store the minimum number of coins in the memo dictionary
        memo.put(amount, Math.min(includeLastCoin, excludeLastCoin));

        // Return the minimum number of coins
        return memo.get(amount);
    }

    public static void main(String[] args) {
        int[] coins = {25, 1, 10, 5};
        int amount = 73;

        int minCoins = makeChangeMinCoins(coins, amount);

        System.out.println("Minimum number of coins to make " + amount + " cents with coins: " + minCoins);
    }
}

File: dsa-23au/java-dsa/pswish-pipeline/Homework_problems/week7/coins.py
def make_change_min_coins(coins, amount, memo={}):
    coins.sort(reverse=True)  # Sort coins in descending order
    if amount == 0:
        return 0
    if amount < 0 or not coins:
        return float('inf')
    if (len(coins), amount) in memo:
        return memo[(len(coins), amount)]

    include_last_coin = 1 + make_change_min_coins(coins, amount - coins[0], memo)  # Use coins[0] as it is now the largest coin
    exclude_last_coin = make_change_min_coins(coins[1:], amount, memo)

    memo[(len(coins), amount)] = min(include_last_coin, exclude_last_coin)
    print(memo)

    return memo[(len(coins), amount)]

# Example usage
coins = [25, 1, 10, 5]
amount = 73
min_coins = make_change_min_coins(coins, amount)
print(f"Minimum number of coins to make {amount} cents with coins: {min_coins}")